File: BOJ/자료구조/1717.cpp
// 1717 집합의 표현
// https://www.acmicpc.net/problem/1717
#include <iostream>
#include <vector>

using namespace std;

int nodes[1000001];

int Find(int x) {
    if (nodes[x] != x) nodes[x] = Find(nodes[x]);
    return nodes[x];
}

void Union(int a, int b) {
    a = Find(a);
    b = Find(b);

    if (a < b) nodes[b] = a;
    else nodes[a] = b;
    return;
}

int main(void) {
    ios::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);

    int N, M;
    
    cin >> N >> M;
    for (int i = 1; i <= N; ++i) nodes[i] = i;
    for (int i = 0; i < M; ++i) {
        int c, a, b;
        
        cin >> c >> a >> b;

        if (c == 0) {
            Union(a, b);
        } else {
            if (Find(a) == Find(b)) {
                cout << "YES\n";
            } else {
                cout << "NO\n";
            }
        }
    }

    return 0;
}

File: BOJ/자료구조/Set/1717/1717.py
# 1717 집합의 표현
# https://www.acmicpc.net/problem/1717
import sys
sys.setrecursionlimit(10**7)
input = sys.stdin.readline


def find(parent, x):
    if parent[x] != x:
        parent[x] = find(parent, parent[x])
    return parent[x]


def union(parent, a, b):
    a = find(parent, a)
    b = find(parent, b)

    if a < b:
        parent[b] = a
    else:
        parent[a] = b


n, m = map(int, input().split())
parentNodes = [i for i in range(n+1)]

for _ in range(m):
    cmd, a, b = map(int, input().split())

    if cmd == 0:
        union(parentNodes, a, b)
    else:
        if find(parentNodes, a) == find(parentNodes, b):
            print("YES")
        else:
            print("NO")
